wefe/debias/double_hard_debias.py

```diff
@@ -0,0 +1,490 @@
+    |      CoRR, abs/2005.00965, 2020.https://arxiv.org/abs/2005.00965.
+    | [2]: https://github.com/uvavision/Double-Hard-Debias
+    """
+
+    name = "Double Hard Debias"
+    short_name = "DHD"
+
+    def __init__(
+        self,
+        pca_args: Dict[str, Any] = {"n_components": 10},
+        verbose: bool = False,
+        criterion_name: Optional[str] = None,
+        incremental_pca: bool = True,
+    ) -> None:
+        """Initialize a Double Hard Debias instance.
+
+        Parameters
+        ----------
+        pca_args : Dict[str, Any], optional
+            Arguments for the PCA that is calculated internally in the identification
+            of the bias subspace, by default {"n_components": 10}
+        verbose : bool, optional
+            True will print informative messages about the debiasing process,
+            by default False.
+        criterion_name : Optional[str], optional
+            The name of the criterion for which the debias is being executed,
+            e.g., 'Gender'. This will indicate the name of the model
+            returning transform, by default None
+        incremental_pca: bool, optional
+            If `True`, incremental pca will be used over the entire set of
+            vectors.
+            If `False`, pca will be used over the entire set of vectors.
+            **WARNING:** Running pca over the entire set of vectors may
+            raise to `MemoryError`,  by default True.
+        """
+        # check verbose
+        if not isinstance(verbose, bool):
+            raise TypeError(f"verbose should be a bool, got {verbose}.")
+
+        # check incremental pca
+        if not isinstance(incremental_pca, bool):
+            raise TypeError(f"incremental_pca should be a bool, got {verbose}.")
+
+        if incremental_pca:
+            self.pca_type = IncrementalPCA()
+        else:
+            self.pca_type = PCA(svd_solver="randomized")
+
+        if not isinstance(pca_args, dict):
+            raise TypeError(f"pca_args should be a dict, got {verbose}.")
+
+        self.pca_args = pca_args
+        self.verbose = verbose
+
+        if criterion_name is None or isinstance(criterion_name, str):
+            self.criterion_name_ = criterion_name
+        else:
+            raise ValueError(f"criterion_name should be str, got: {criterion_name}")
+
+    def _check_sets_size(
+        self, sets: Sequence[Sequence[str]], set_name: str,
+    ):
+
+        for idx, set_ in enumerate(sets):
+            if len(set_) != 2:
+                adverb = "less" if len(set_) < 2 else "more"
+
+                raise ValueError(
+                    f"The {set_name} pair at position {idx} ({set_}) has {adverb} "
+                    f"words than allowed by {self.name}: "
+                    f"got {len(set_)} words, expected 2."
+                )
+
+    def _similarity(self, u: List[np.ndarray], v: List[np.ndarray]) -> float:
+        return 1 - pairwise_distances(u, v, metric="cosine")
```

Comment:
The return rate is bad. 
I think it should be np.ndarray.

Reply:
fixed